python/ray/util/spark/cluster_init.py

```diff
@@ -529,14 +540,50 @@ def _setup_ray_cluster(
 
     cluster_unique_id = uuid.uuid4().hex[:8]
 
-    if ray_temp_root_dir is None:
-        ray_temp_root_dir = start_hook.get_default_temp_dir()
-    ray_temp_dir = os.path.join(
-        ray_temp_root_dir, f"ray-{ray_head_port}-{cluster_unique_id}"
-    )
-    os.makedirs(ray_temp_dir, exist_ok=True)
-    object_spilling_dir = os.path.join(ray_temp_dir, "spill")
-    os.makedirs(object_spilling_dir, exist_ok=True)
+    if os.environ.get("RAY_ON_SPARK_GLOBAL_MODE", "0") == "1":
+        # global mode enabled
+        # for global mode, Ray always uses default temp dir
+        # so that local Ray client can discover it without specifying
+        # head node address.
+        if ray_temp_root_dir is not None:
+            raise ValueError(
+                "Ray on spark global mode cluster does not allow you to set "
+                "'ray_temp_root_dir' argument."
+            )
+
+        # We only allow user to launch one active Ray on spark global cluster
+        # at a time. So acquiring a global file lock before setting up a new
+        # Ray on spark global cluster.
+        global_cluster_lock_fd = os.open(
+            "/tmp/ray_on_spark_global_cluster.lock",
+            os.O_RDWR | os.O_CREAT | os.O_TRUNC
+        )
+
+        try:
+            # acquiring exclusive lock to ensure copy logs and removing dir safely.
+            fcntl.flock(global_cluster_lock_fd, fcntl.LOCK_EX | fcntl.LOCK_NB)
+        except BlockingIOError:
+            # acquiring global lock failed.
+            raise ValueError(
+                "Acquiring global lock failed for setting up new global mode Ray on "
+                "spark cluster. If there is an active global mode Ray on spark "
+                "cluster, please shut down it before you create a new one."
+            )
+
+        ray_temp_dir = None
+        root_tmp_dir = os.environ.get("RAY_TMPDIR", "/tmp")
```

Comment:
Should we use `start_hook.get_default_temp_root_dir()` as below?

Reply:
If we don't set `--ray-temp-dir`, it will create tmp dir under $RAY_TMPDIR , and I prefer to put spill directory under created ray tmp dir